File: JavaScript/hw9/ConvexPolygon.js
class ConvexPolygon extends Shape{
    constructor(v){
        super(v);
        this.name = 'ConvexPolygon';
    }
    description(){
        var s = '{' + this.name + ';'
        for(var i = 0; i < this.n; i++){
            s += '(' + this.vertices[i].component({'index':1}).toString();
            s += ',' + this.vertices[i].component({'index':2}).toString();
            s += ')/';
        }
        return s.substring(0,s.length - 1) + '}';
    }
    area(){
        var sum = 0;
        var v = this.vertices;
        for (var i = 0; i < this.n - 1; i++)
            sum += v[i].component()[0] * v[i+1].component()[1] - v[i].component()[1] * v[i+1].component()[0];
        return Math.abs(sum)/2;
    }
    perimeter() {
        var p = 0;
        for (var i = 0; i < this.n-1; i++) {
            p += Vector.sub(this.vertices[i], this.vertices[i+1]).length();
        }
        p += Vector.sub(this.vertices[0], this.vertices[this.n-1]).length();
        return p;
    }
}

function createConvexPolygon(s, separator = '/'){
    s = s.split(separator);
    var v = []
    for(var i in s){
        if(s[i].charAt(0) != '(' || s[i].charAt(s[i].length - 1) != ')')
            throw new Error('Format Error');
        s[i] = s[i].substring(1,s[i].length - 1);
        s[i] = s[i].split(',');
        v.push( new Vector( [ parseInt(s[i][0]), parseInt(s[i][1]) ] ) );
    }
    sortVector(v);
    return new ConvexPolygon(v);
}

//           ffffffff
//          fffffffff
//         ffff
//         ffff
//         ffff
// ffffffffffffffffffff     uuu         uuu
// ffffffffffffffffffff     uuu         uuu
//         ffff             uuu         uuu
//         ffff             uuu         uuu
//         ffff              uuu       uuuuu
//         ffff               uuu     uuu uu
//         ffff                 uuuuuuu    uuu

File: Python/hw9/ConvexPolygon.py
import math
import Vector
import VectorInput
import Shape


class ConvexPolygon(Shape.Shape):
    def __init__(self, v):
        self.n = len(v)
        self.v = []
        for i in v:
            self.v.append(i)

    def verticesNum(self):
        return self.n

    def vertices(self, n = None):
        if n is None:
            l = []
            for i in self.v:
                l.append(i.comp)
            return l
        else:
            return self.v[n]

    def perimeter(self):
        i = -1
        p = 0
        while i < self.n - 1:
            p += (self.v[i+1]-self.v[i]).length()
            i += 1
        return p

    def area(self):
        o = Vector.centroid(self.v)
        i = -1
        result = 0
        while i < self.n - 1:
            a = (o - self.v[i]).length()
            b = (o - self.v[i+1]).length()
            c = (self.v[i+1] - self.v[i]).length()
            s = (a + b + c) / 2
            result += math.sqrt(s * (s - a) * (s - b) * (s - c))
            i += 1
        return result

    def description(self):
        s = '{convex polygon, '
        s += str(self.n) + ', '
        for i in range(0, len(self.v)):
            s += '('
            for j in range(1, self.v[i].dimension()):
                s += str(self.v[i].component(j)) + ','
            s += str(self.v[i].comp[-1]) + ')/'
        s = s[0:-1]
        s += '}'
        return s


def createConvexPolygon(s):
    v = VectorInput.getVectorList(s)
    if len(v) < 3:
        raise Exception("Less than 3 vertices")
    Vector.sortVector(v)
    return ConvexPolygon(v)
